Map LDAP account attributes to labels such as groups #63

kcd83 opened the pull request:
**Updated:**

I am proposing a configurable structure to leverage ldap account attributes as labels for access control. Let me know what you think.

I have added a `labels` mapping to the configuration, for example:
```
ldap_auth:
  ...
  filter: (&(|(userPrincipalName=${account})(userPrincipalName=${account}@example.com)(mail=${account}))(objectClass=person))
  labels:
    groups:
      attribute: memberOf
      parse_cn: true
    title:
      attribute: title
```

Can then be used with the current labels implementation like so:

```
 acl:
  - match: {labels: {"title": "Developer"}}
    actions: ["*"]
    comment: "If you call yourself a developer you can do anything."
  - match: {labels: {"groups": "Admin"}}
    actions: ["push"]
    comment: "If you are part of the admin group you can push."	
```
	
Using title for an acl is probably a really dumb idea. But it proves you can map arbitary attributes to arbitary labels. 

There is one special option, `parse_cn` which extracts the common name from a fully qualified value as found in `memberOf`.


**Original:**

I know little about LDAP and even less GO but I've taken the labels approach in #139 and populated groups to address #63.

```
  - match: {name: "docker_auth", labels: {"groups": "Admin"}}
    actions: ["*"]
    comment: "Admin users can do anything to docker_auth."
```

I'm simply looking through the account response for `memberOf` details. The alternative approach for ldap would be to look up groups and find member accounts.

A similar approach might be useful for #117

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cesanta/docker_auth/167)
<!-- Reviewable:end -->

rojer commented:
@kcd83 thanks for the contribution! this looks very reasonable to me, but then again - i know very little about LDAP too.
have you actually used it yourself? if not, i'd like someone who actually uses LDAP to confirm that this works.

maybe @summerQLin can comment

kcd83 commented:
Apparently modern AD installations/ configurations don't necessarily expose `memberof` and/or may prefer impersonation... anyone with ldap experience able to comment?

mrueg commented:
Is there any chance to merge this soon and split the AD part into a separate issue?

alonbl commented:
Hi,

I checked this on top of master and I can confirm it is working!

It is usable as-is so please merge it.

The problem with this approach is that the memberOf resolution does not support recursive groups. However, even with immediate groups there is a value.

Validated with AD.

I do need recursive groups so I ended up with the following filter instead of using this patch:
````
    (&(sAMAccountName=${account})(objectClass=person)(memberOf:1.2.840.113556.1.4.1941:=CN=XXX,OU=YYY,DC=X))
````
This is not flexible as I wish it to be... as I can use only single recursive group.

It is hard for me to imagine how in this generic attribute -> label approach a recursive solution may be built without code change. For active directory a list of filters to check recursive each group provides a resolution, for example the following will set in_group1 to "group1" if user is recursively in group1.

```
  filters:
    base:
      filter: <this is the base filter as currently implemented
    group1:
      filter: (&(sAMAccountName=${account})(objectClass=person)(memberOf:1.2.840.113556.1.4.1941:=CN=group1,OU=YYY,DC=X))
      labels:
        in_group1:
          attribute: cn
```
In other directories it is hard to perform this without recursion. As groups are so important, maybe a special behaviour is needed.

Thanks!

alonbl commented on auth_server/authn/ldap_auth.go:
```diff
@@ -27,16 +26,20 @@ import (
+	Filter                string              `yaml:"filter,omitempty"`
+	BindDN                string              `yaml:"bind_dn,omitempty"`
+	BindPasswordFile      string              `yaml:"bind_password_file,omitempty"`
+	LabelMaps             map[string]LabelMap `yaml:"labels,omitempty"`
```
please add doc into examples/reference.yml otherwise nobody will know :)

pinepain replied:
It seems to be this is already added to `examples/reference.yml`

mrueg commented:
@kcd83 any chance to update it to @alonbl's comments so it could be merged? :)

kcd83 commented:
@mrueg @alonbl I have added the example configuration to reference.yml

I agree that additional queries of groups would be great but it isn't something I'll be able to look at. Mostly because this is all the Go I've ever written and I could barely remember how to build it :)

alonbl commented on examples/reference.yml:
```diff
@@ -134,6 +134,16 @@ ldap_auth:
+    groups:
+      attribute: memberOf
+      # Special handling to simplify the values to just the common name
+      parse_cn: true
```
Good, I would also suggest to provide an example at acl section of how to use these attributes

pinepain replied:
For those who are not very familiar with LDAP, it would be nice to provide example when `parse_cn` should be `true` and when `false` (false is default behavior, right)?

kcd83 replied:
It is a special case _only_ for `cn` fields, it isn't very reusable

pinepain replied:
That's just to get `foo` from `cn=foo,ou=bar,dc=bar`, right?

alonbl commented:
Thanks @kcd83! As I wrote, this patch is helpful as-is, the group thing can be added later.

rojer commented:
please address @alonbl 's comments about examples and docs and this can be merged. thanks!

kcd83 commented:
Whoops i was supposed to call that group not groups to match the existing acls. Would it be worth saying "see the group acls below" or similar also?

```
 - match: {labels: {"group": "VIP"}}
    actions: ["push"]
    comment: "Users assigned to group 'VIP' is able to push"
  - match: {labels: {"group": "/trainee|dev/"}}
    actions: ["push", "pull"]
    comment: "Users assigned to group 'trainee' and 'dev' is able to push and pull"
 

```

alonbl commented:
You gave a great example in your comment#0 :)
I tested based on that...

```
acl:
  - match: {labels: {"title": "Developer"}}
    actions: ["*"]
    comment: "If you call yourself a developer you can do anything."
  - match: {labels: {"groups": "Admin"}}
    actions: ["push"]
    comment: "If you are part of the admin group you can push."	
```

yes, a nice comment will be helpful as well.

pinepain commented:
@kcd83 Thank you for this nice feature! 

Are there any estimates on when this could be good to merge and released? 

My use case is to authenticate and authorize users with [JumpCloud LDAP](https://support.jumpcloud.com/customer/en/portal/articles/2439911), the [Jenkins Integration Guide](https://support.jumpcloud.com/customer/en/portal/articles/2440013-configuring-jenkins-to-use-jumpcloud-s-ldap-as-a-service) has some examples to filter by user groups

kcd83 commented:
I _think_ I addressed all the suggestions.

I don't think this is the place for an LDAP tutorial however, particularly because it is a very limited implementation. The examples can be used and someone with more ldap knowledge could improve this in time. 

Ok to merge?

pcallewaert commented:
Hi,

This probably has no support for RFC 2307 style where group membership has members listed on the group object rather than using groups listed on user object?

pinepain commented:
@kcd83 Thanks for you hard work! Can't wait to have this changes in a release. Would you have some time to rebase your branch onto current master so that there will be no conflicts and to make it easier for @rojer to merge it by just pressing the merge button?

@rojer do you think this PR is good to be merged?

mrueg commented:
@kcd83 can you update this PR to remove the file conflicts?
@rojer I've been using this patch for quite a while, it works well. Can we get it merged?

aparamon commented:
This feature is desirable -- any obstacle to merge?

rojer commented:
closing in favor of https://github.com/cesanta/docker_auth/pull/228
